File: frontend/test/duel_score_test.dart
import 'package:srbski_read/course/duel_score.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('удар за предложение', () {
    test('выигранное предложение бьёт по машине и не задевает вас', () {
      final blow = blowFor(DuelWinner.user, 60);
      expect(blow.toFoe, winDamage);
      expect(blow.toHero, 0);
    });

    test('проигранное бьёт по вам и не задевает машину', () {
      final blow = blowFor(DuelWinner.translator, 100);
      expect(blow.toHero, winDamage);
      expect(blow.toFoe, 0);
    });

    test('ничья задевает обоих поровну', () {
      final blow = blowFor(DuelWinner.tie, 50);
      expect(blow.toFoe, blow.toHero);
      expect(blow.toFoe, greaterThan(0));
    });

    // Самое важное свойство всей математики: разгон не имеет доступа к
    // полоскам. Пока он их двигал, быстрый набор отдавал раунд человеку,
    // проигравшему по счёту три предложения из пяти.
    test('разгон не двигает полоски здоровья', () {
      for (final winner in DuelWinner.values) {
        final slow = blowFor(winner, 0);
        final fast = blowFor(winner, 100);
        expect(fast.toFoe, slow.toFoe);
        expect(fast.toHero, slow.toHero);
      }
    });

    test('разгон поднимает очки стиля', () {
      expect(blowFor(DuelWinner.user, 100).style,
          greaterThan(blowFor(DuelWinner.user, 0).style));
    });

    test('даже без разгона выигранное предложение даёт стиль', () {
      expect(blowFor(DuelWinner.user, 0).style, greaterThan(0));
    });

    test('проигранное предложение стиля не даёт', () {
      expect(blowFor(DuelWinner.translator, 100).style, 0);
    });

    test('критом считается только удар с полным разгоном', () {
      expect(blowFor(DuelWinner.user, critPower - 1).crit, isFalse);
      expect(blowFor(DuelWinner.user, critPower).crit, isTrue);
    });

    test('множитель разгона зажат между полом и потолком', () {
      expect(powerFactor(-50), powerFactor(0));
      expect(powerFactor(500), powerFactor(100));
      expect(powerFactor(double.nan), powerFactor(0));
    });
  });

  // Полоски — это счёт. Ниже перебираются все расклады раунда из пяти фраз и
  // проверяется, что итог по полоскам совпадает с итогом по числу побед: те же
  // 243 сочетания, что и в веб-тесте (web/src/lib/duelScore.test.ts).
  group('итог раунда повторяет счёт', () {
    DuelWinner play(List<DuelWinner> winners, double power) {
      var hero = roundHp;
      var foe = roundHp;
      for (final winner in winners) {
        final blow = blowFor(winner, power);
        hero = (hero - blow.toHero).clamp(0, roundHp);
        foe = (foe - blow.toFoe).clamp(0, roundHp);
      }
      return roundOutcome(hero, foe);
    }

    DuelWinner byCount(List<DuelWinner> winners) {
      final wins = winners.where((item) => item == DuelWinner.user).length;
      final losses =
          winners.where((item) => item == DuelWinner.translator).length;
      if (wins == losses) return DuelWinner.tie;
      return wins > losses ? DuelWinner.user : DuelWinner.translator;
    }

    test('во всех 243 раскладах и при любом разгоне', () {
      final every = <List<DuelWinner>>[];
      void build(List<DuelWinner> prefix) {
        if (prefix.length == 5) {
          every.add(prefix);
          return;
        }
        for (final winner in DuelWinner.values) {
          build([...prefix, winner]);
        }
      }

      build([]);
      expect(every.length, 243);

      for (final winners in every) {
        for (final power in [0.0, 50.0, 100.0]) {
          expect(play(winners, power), byCount(winners),
              reason: '${winners.join(',')} при разгоне $power');
        }
      }
    });

    test('чистая победа кладёт машину в ноль', () {
      var foe = roundHp;
      for (var i = 0; i < 5; i++) {
        foe -= blowFor(DuelWinner.user, 0).toFoe;
      }
      expect(foe, 0);
    });
  });

  group('серия', () {
    test('ступень растёт с серией и упирается в число файлов клика', () {
      expect(comboStep(0), 0);
      expect(comboStep(1), 0);
      expect(comboStep(4), 2);
      expect(comboStep(64), 6);
      expect(comboStep(100000), 7);
    });

    test('кричит каждые восемь символов', () {
      expect(isComboMilestone(8), isTrue);
      expect(isComboMilestone(16), isTrue);
      expect(isComboMilestone(9), isFalse);
      expect(isComboMilestone(0), isFalse);
    });
  });

  group('ранг', () {
    test('раздаётся по очкам стиля', () {
      expect(rankOf(140), 'S');
      expect(rankOf(110), 'A');
      expect(rankOf(70), 'B');
      expect(rankOf(20), 'C');
    });

    test('пять выигранных фраз без пауз дают высший ранг, с раздумьями — нет', () {
      expect(rankOf(5 * blowFor(DuelWinner.user, 100).style), 'S');
      expect(rankOf(5 * blowFor(DuelWinner.user, 0).style), 'B');
    });
  });
}

File: web/src/lib/duelScore.test.ts
import { describe, expect, it } from 'vitest';

import {
  blowFor,
  comboStep,
  isComboMilestone,
  powerFactor,
  rankOf,
  roundOutcome,
  CRIT_POWER,
  ROUND_HP,
  WIN_DAMAGE,
  type DuelWinner,
} from './duelScore';

describe('удар за предложение', () => {
  it('выигранное предложение бьёт по машине и не задевает вас', () => {
    const blow = blowFor('user', 60);
    expect(blow.toFoe).toBe(WIN_DAMAGE);
    expect(blow.toHero).toBe(0);
  });

  it('проигранное бьёт по вам и не задевает машину', () => {
    const blow = blowFor('translator', 100);
    expect(blow.toHero).toBe(WIN_DAMAGE);
    expect(blow.toFoe).toBe(0);
  });

  it('ничья задевает обоих поровну', () => {
    const blow = blowFor('tie', 50);
    expect(blow.toFoe).toBe(blow.toHero);
    expect(blow.toFoe).toBeGreaterThan(0);
  });

  /*
    Самое важное свойство всей математики: разгон не имеет доступа к полоскам.
    Пока он их двигал, быстрый набор отдавал раунд человеку, проигравшему по
    счёту три предложения из пяти.
  */
  it('разгон не двигает полоски здоровья', () => {
    for (const winner of ['user', 'translator', 'tie'] as DuelWinner[]) {
      const slow = blowFor(winner, 0);
      const fast = blowFor(winner, 100);
      expect(fast.toFoe).toBe(slow.toFoe);
      expect(fast.toHero).toBe(slow.toHero);
    }
  });

  it('разгон поднимает очки стиля', () => {
    expect(blowFor('user', 100).style).toBeGreaterThan(blowFor('user', 0).style);
  });

  it('даже без разгона выигранное предложение даёт стиль', () => {
    // Перевод, обдуманный медленно, — всё равно выигранный перевод.
    expect(blowFor('user', 0).style).toBeGreaterThan(0);
  });

  it('проигранное предложение стиля не даёт', () => {
    expect(blowFor('translator', 100).style).toBe(0);
  });

  it('критом считается только удар с полным разгоном', () => {
    expect(blowFor('user', CRIT_POWER - 1).crit).toBe(false);
    expect(blowFor('user', CRIT_POWER).crit).toBe(true);
  });

  it('пропущенный удар не бывает критическим', () => {
    expect(blowFor('translator', 100).crit).toBe(false);
  });

  it('множитель разгона зажат между полом и потолком', () => {
    expect(powerFactor(-50)).toBe(powerFactor(0));
    expect(powerFactor(500)).toBe(powerFactor(100));
    expect(powerFactor(Number.NaN)).toBe(powerFactor(0));
  });
});

/*
  Полоски — это счёт. Ниже перебираются все расклады раунда из пяти фраз и
  проверяется, что итог по полоскам совпадает с итогом по числу побед. Ровно
  этого и не хватало: зрелище обязано повторять счёт, а не спорить с ним.
*/
describe('итог раунда повторяет счёт', () => {
  const play = (winners: DuelWinner[], power: number) => {
    let hero = ROUND_HP;
    let foe = ROUND_HP;
    for (const winner of winners) {
      const blow = blowFor(winner, power);
      hero = Math.max(0, hero - blow.toHero);
      foe = Math.max(0, foe - blow.toFoe);
    }
    return { outcome: roundOutcome(hero, foe), hero, foe };
  };

  const byCount = (winners: DuelWinner[]): DuelWinner => {
    const wins = winners.filter((winner) => winner === 'user').length;
    const losses = winners.filter((winner) => winner === 'translator').length;
    if (wins === losses) return 'tie';
    return wins > losses ? 'user' : 'translator';
  };

  const every: DuelWinner[][] = [];
  const build = (prefix: DuelWinner[]) => {
    if (prefix.length === 5) {
      every.push(prefix);
      return;
    }
    for (const winner of ['user', 'translator', 'tie'] as DuelWinner[]) build([...prefix, winner]);
  };
  build([]);

  it('во всех 243 раскладах и при любом разгоне', () => {
    for (const winners of every) {
      for (const power of [0, 50, 100]) {
        expect(play(winners, power).outcome, `${winners.join(',')} при разгоне ${power}`).toBe(
          byCount(winners),
        );
      }
    }
  });

  it('чистая победа кладёт машину в ноль', () => {
    expect(play(['user', 'user', 'user', 'user', 'user'], 0).foe).toBe(0);
  });

  it('чистое поражение кладёт вас в ноль', () => {
    expect(play(['translator', 'translator', 'translator', 'translator', 'translator'], 100).hero).toBe(0);
  });
});

describe('серия', () => {
  it('ступень растёт с серией и упирается в число файлов клика', () => {
    expect(comboStep(0)).toBe(0);
    expect(comboStep(1)).toBe(0);
    expect(comboStep(4)).toBe(2);
    expect(comboStep(64)).toBe(6);
    expect(comboStep(100000)).toBe(7);
  });

  it('кричит каждые восемь символов', () => {
    expect(isComboMilestone(8)).toBe(true);
    expect(isComboMilestone(16)).toBe(true);
    expect(isComboMilestone(9)).toBe(false);
    expect(isComboMilestone(0)).toBe(false);
  });
});

describe('ранг', () => {
  it('раздаётся по очкам стиля', () => {
    expect(rankOf(140)).toBe('S');
    expect(rankOf(110)).toBe('A');
    expect(rankOf(70)).toBe('B');
    expect(rankOf(20)).toBe('C');
  });

  it('пять выигранных фраз без пауз дают высший ранг, с раздумьями — нет', () => {
    const fast = 5 * blowFor('user', 100).style;
    const slow = 5 * blowFor('user', 0).style;
    expect(rankOf(fast)).toBe('S');
    expect(rankOf(slow)).toBe('B');
  });
});
